packages/react-instantsearch/src/connectors/connectMenu.js

```diff
@@ -77,19 +77,37 @@ export default createConnector({
       return null;
     }
 
-    const items = results
-      .getFacetValues(attributeName, {sortBy})
-      .slice(0, limit)
-      .map(v => ({
-        value: getValue(v.name, props, searchState),
-        label: v.name,
-        count: v.count,
-        isRefined: v.isRefined,
-      }));
+    const items = searchForFacetValuesResults && searchForFacetValuesResults[attributeName]
+      ? searchForFacetValuesResults[attributeName]
+        .map(
+          v => ({
+            label: v.value,
+            value: getValue(v.value, props, searchState),
+            _highlightResult: {label: {value: v.highlighted}},
+            count: v.count,
+            isRefined: v.isRefined,
+          }))
+      : results
+        .getFacetValues(attributeName, {sortBy})
+        .map(v => ({
+          label: v.name,
+          value: getValue(v.name, props, searchState),
+          count: v.count,
+          isRefined: v.isRefined,
+        }));
+
+    const isFromSearch = Boolean(searchForFacetValuesResults && searchForFacetValuesResults[attributeName]);
+    const searchForFacetValues = props.searchForFacetValues ? this.searchForFacetValues : undefined;
+
+    const sortedItems = !isFromSearch && props.searchForFacetValues ?
+      orderBy(items, ['isRefined', 'count', 'label'], ['desc', 'desc', 'asc']) : items;
+    const transformedItems = props.transformItems ? props.transformItems(sortedItems) : sortedItems;
 
     return {
-      items,
+      items: transformedItems.slice(0, limit),
```

Comment:
After second thought, the behaviour you implemented is the right one because when being given items to reorder, you do need to loop through ALL of them before being able to display only ten of them so we are safe in doing this slice AFTER transformItems